src/levels.js

```diff
@@ -1,91 +1,111 @@
+const defaults = require('./defaults');
+const text = require('./text');
 
 /**
- * Normalize the array
- * @param arr
- * @param opts
- * @returns {*}
+ * Check whether n is a valid logger number.
+ * @param n object to be checked
+ * @returns boolean
+ */
+const isNumberValid = (n) => Number.isInteger(parseFloat(n)) && _.inRange(n, 8);
+
+/**
+ * Normalize logger levels array.
+ * Checks number of levels and validity of each level.
+ * @param {Array} arr array containing logger levels
+ * @param {Map} opts original options object - used for output in case of error
+ * @returns {Array}
  */
 const normArr = (arr, opts) => {
   if (arr.length > 8) {
     throw new BadOptionsError(opts, text.tooManyLevels(arr.length));
   }
 
-  return arr.map(val => {
-    if (val && _.isString(val)) return val;
-    if (_.isNumber(val) && isFinite(val)) return val.toString();
-    if (_.isNull(val) || _.isUndefined(val)) return undefined;
+  return arr.map((val) => {
+    if (val && _.isString(val)) {return val;}
+    if (_.isNumber(val) && Number.isFinite(val)) {return val.toString();}
+    if (_.isNull(val) || _.isUndefined(val)) {return undefined;}
 
     throw new BadOptionsError(opts, text.levelNotString(val));
   });
 };
 
 
 /**
- * Normalize the object
- * @param obj
- * @param opts
+ * Normalize logger levels dictionary.
+ * Check the values dictionary for duplicates and invalid levels
+ * @param obj dictionary containing levels
+ * @param opts original options object - used for output in case of error
  */
 const normObj = (obj, opts) => {
   const lvlNums = _.values(obj);
 
-  for (const num of lvlNums) {
+  lvlNums.forEach((num) => {
     if (!isNumberValid(num)) {
       throw new BadOptionsError(opts, text.invalidLevelNum(num));
     }
-  }
+  });
 
-  const duplicates =
-      _(obj).countBy().pick(lvl => lvl > 1)
-          .keys()
-          .value();
+  //  Check for duplicates
+  const duplicates = _(obj).countBy().pick((lvl) => lvl > 1)
+    .keys()
+    .value();
 
-  if (duplicates.length) {
+  if (duplicates.length > 0) {
     throw new BadOptionsError(opts, text.duplicateLevelNums(duplicates));
   }
 
   return _.reduce(obj, (arr, i, name) => {
     const reducedArr = arr;
+
     reducedArr[i] = name;
+
     return reducedArr;
   }, []);
 };
 
 /**
- * Main normalize function to export.
- * @param opts
- * @returns {*}
+ * Takes an options dictionary, normalizes embedded levels if they exist
+ * otherwise returns defaults.
+ * @param opts options dictionary
+ * @returns array
  */
-export const normalize = (opts) => {
+const normalize = (opts) => {
   let custom = opts.levels;
 
+  //  Ensure custom is object
   if (!_.isUndefined(custom) && !_.isNull(custom) && !_.isObject(custom)) {
     throw new BadOptionsError(opts, text.levelsNotObj(typeof custom));
   }
 
+  //  If no custom options are provided, return defaults
   if (!custom) {
-    return defaults.levels.slice();
+    return defaults.levels;
```

Comment:
```suggestion
    return defaults.levels.slice();
```
The `.slice()` was probably intentional as a copy operation, as it comes from defaults and we don't want to pass a reference to our defaults to outside the library.

Reply:
You're right, accidentally left this in when I was trying to re-write the library to use dict rather than arrays. added back in